contracts/src/Summa.sol

```diff
@@ -9,163 +9,128 @@ import "@openzeppelin/contracts/access/Ownable.sol";
-    mapping(bytes32 => IBalanceRetriever) public balanceRetriever;
-    //A mapping of address type hash to the address ownership verifier smart contract for that type of address
-    mapping(bytes32 => IAddressOwnershipVerifier) public cexAddressVerifier;
-
     /**
-     * @dev Struct representing an address owned by the CEX
-     * @param addressType The type of the address by chain (e.g., EVM, BTC)
-     * @param cexAddress The address owned by the CEX
-     * @param ownershipProof The data needed to verify the address ownership (e.g., in case of EVM, a signature and a message signed by the address, encoded as bytes)
+     * @dev Struct representing an address ownership proof submitted by the CEX
+     * @param cexAddress The address owned by the CEX (submitted as a string, as it can be a non-EVM address)
+     * @param chain The name of the chain name where the address belongs (e.g., ETH, BTC)
+     * @param signature The signature of the message signed by the address public key
+     * @param message The message signed by the address public key
      */
-    struct OwnedAddress {
-        bytes32 addressType;
-        bytes cexAddress;
-        bytes ownershipProof;
+    struct AddressOwnershipProof {
+        string cexAddress;
+        string chain;
+        bytes signature;
+        bytes message;
     }
 
     /**
      * @dev Struct representing an asset owned by the CEX
-     * @param assetType The type of the asset (e.g., a hash of the asset kind, like keccak256("ETH"), keccak256("BTC", keccak256("ERC20")
-     * @param amountToProve The amount of the asset that the CEX wants to prove to own
-     * @param addresses CEX addresses that control the @param asset
-     * @param balanceRetrieverArgs Additional arguments needed to get the balance (e.g., in case of ERC20, the address of the ERC20 contract)
+     * @param assetName The name of the asset
+     * @param chain The name of the chain name where the asset lives (e.g., ETH, BTC)
      */
-    struct OwnedAsset {
-        bytes32 assetType;
-        uint256 amountToProve;
-        bytes[] addresses;
-        bytes balanceRetrieverArgs;
+    struct Asset {
+        string assetName;
+        string chain;
+        uint256 amount;
     }
 
-    event AddressVerifierSet(bytes32 indexed addressType, address verifier);
-    event BalanceRetrieverSet(
-        bytes32 indexed assetType,
-        address balanceRetriever
+    /**
+     * @dev Zero-knowledge proof and its inputs
+     * @param proof zero-knowledge proof
+     * @param publicInputs proof inputs
+     */
+    struct ZKProof {
+        bytes proof;
+        uint256[] publicInputs;
+    }
+
+    IVerifier private immutable verifier;
+
+    //All address ownership proofs submitted by the CEX
+    AddressOwnershipProof[] public addressOwnershipProofs;
+
+    //Convenience mapping to check if an address has already been verified
+    mapping(bytes32 => AddressOwnershipProof) public ownershipProofByAddress;
```

Comment:
How about using `i`, index of `addressOwnershipProofs` instead of `AddressOwnershipProof` in here?

Reply:
I wanted to use the address (or its hash) for convenience. First, there's a check whether this particular address has previously been verified (line 84), so the check should perform some kind of lookup by address. Second, even in case of manual querying by a human, they would probably be searching for the proof of ownership of a particular address, so the mapping allows to do that in a single query (instead of iterating over the array).

Reply:
Actually a good idea! Will change.